legacy/builder/phases/linker.go

```diff
@@ -63,46 +65,150 @@ func (s *Linker) Run(ctx *types.Context) error {
 	return nil
 }
 
+// CppArchive represents a cpp archive (.a). It has a list of object files
+// that must be part of the archive and has functions to build the archive
+// and check if the archive is up-to-date.
+type CppArchive struct {
+	ArchivePath   *paths.Path
+	CacheFilePath *paths.Path
+	Objects       paths.PathList
+}
+
+// NewCppArchive creates an empty CppArchive
+func NewCppArchive(archivePath *paths.Path) *CppArchive {
+	return &CppArchive{
+		ArchivePath:   archivePath,
+		CacheFilePath: archivePath.Parent().Join(archivePath.Base() + ".cache"),
+		Objects:       paths.NewPathList(),
+	}
+}
+
+// AddObject adds an object file in the list of files to be archived
+func (a *CppArchive) AddObject(object *paths.Path) {
+	a.Objects.Add(object)
+}
+
+func (a *CppArchive) readCachedFilesList() paths.PathList {
+	var cache paths.PathList
+	if cacheData, err := a.CacheFilePath.ReadFile(); err != nil {
+		return nil
+	} else if err := json.Unmarshal(cacheData, &cache); err != nil {
+		return nil
+	} else {
+		return cache
+	}
+}
+
+func (a *CppArchive) writeCachedFilesList() error {
+	if cacheData, err := json.Marshal(a.Objects); err != nil {
+		panic(err) // should never happen
```

Comment:
Shouldn't be better to return an error instead of panicking?

Reply:
In theory marshaling `a.Object` should never fail and we could ignore the error. Panicking here is like asserting that `err == nil`.